File: src/year2015/day03/part1.cc
#include <set>
#include <string>

#include "../solutions.h"
#include "common.h"
#include "parse.h"

auto Year2015::Day03::Part1(const std::string& input) -> std::string {
  const auto directions = parse(input);
  Coordinate position = {.x = 0, .y = 0};
  std::set<Coordinate> houses = {position};
  for (const auto& direction : directions) {
    switch (direction) {
      case Direction::North: {
        position.y++;
        houses.insert(position);
        break;
      }
      case Direction::South: {
        position.y--;
        houses.insert(position);
        break;
      }
      case Direction::East: {
        position.x++;
        houses.insert(position);
        break;
      }
      case Direction::West: {
        position.x--;
        houses.insert(position);
        break;
      }
    }
  }
  return std::to_string(houses.size());
}

File: src/year2015/day03/part1.rs
use std::collections::HashSet;

use super::common::Coordinate;
use super::parse::{Direction, parse};
pub fn part1(input: String) -> Result<String, String> {
  let directions = parse(input);
  if directions.is_err() {
    return Err(directions.unwrap_err());
  }
  let directions = directions.unwrap();
  let mut santa_pos = Coordinate { x: 0, y: 0 };
  let mut positions: HashSet<Coordinate> = HashSet::new();
  positions.insert(santa_pos);

  for direction in directions {
    match direction {
      Direction::North => santa_pos.y += 1,
      Direction::South => santa_pos.y -= 1,
      Direction::East => santa_pos.x += 1,
      Direction::West => santa_pos.x -= 1,
    }
    positions.insert(santa_pos);
  }
  return Ok(positions.len().to_string());
}
